### PR TITLE
U4-10681 Restore option in media recycle bin

### PR DESCRIPTION
**Pull request by richarth**
Added an option to restore media from the media recycle bin

### CODE REVIEW

**Comment by nul800sebastiaan** [comment]
Hey there and thanks for this! 

We don't really want to make any database changes in patch versions, only in minor/major versions. Could you please re-target this PR for 7.8.0 instead of 7.7.7? After that we can review it. Thanks!

**Comment by richarth** [comment]
No problem.  I've now retargeted for 7.8.0.

**Comment by richarth** [comment]
I've now retargeted the changes for 7.9.0

**Comment by richarth** [comment]
I've retargeted my changes for 7.10.  Hopefully third time lucky ;)

**Comment by richarth** [comment]
Now retargeted for 7.11

**Comment by richarth** [comment]
7.12 is now targeted

**Comment by dawoe** [comment]
Hi @richarth,

First of all thx for these PR and especially for keeping it up to date. I will give this a spin this week to see if I can find issues with it. At a first glance the code looks okay.

Dave [Community Pull Request team](https://umbraco.com/blog/introducing-the-new-cpr-team/)

File: src/Umbraco.Core/Strategies/RelateOnTrashHandler.cs [hunk 1]
```diff
@@ -72,5 +89,46 @@ private void ContentService_Trashed(IContentService sender, MoveEventArgs<IConte
...
+                //may not be the case if the encompassing transaction also deleted it when this item was moved to the bin
+
+                if (entityService.Exists(originalParentId))
+                {
+                    // Add a relation for the item being deleted, so that we can know the original parent for if we need to restore later
+                    var relation = new Relation(originalParentId, item.Entity.Id, relationType);
+                    relationService.Save(relation);
+
+                    ApplicationContext.Current.Services.AuditService.Add(AuditType.Delete,
+                        string.Format("Trashed media item with Id: '{0}' related to original parent media item with Id: '{1}'", item.Entity.Id, originalParentId),
```
Comment (dawoe) [inline src/Umbraco.Core/Strategies/RelateOnTrashHandler.cs]: Can you move this text to the language files and use the localization service to get it?

File: src/Umbraco.Web.UI.Client/src/views/media/media.restore.controller.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	    $scope.success = false;
+
+		relationResource.getByChildId(node.id, "relateParentDocumentOnDelete").then(function (data) {
+
+            if (data.length == 0) {
+                $scope.success = false;
+                $scope.error = {
+                    errorMsg: "Cannot automatically restore this item",
+                    data: {
+                        Message: "There is no 'restore' relation found for this node. Use the Move menu item to move it manually."
```
Comment (dawoe) [inline src/Umbraco.Web.UI.Client/src/views/media/media.restore.controller.js]: These texts should be moved to a language file and use the angular localization service to get them

File: src/Umbraco.Web.UI.Client/src/views/media/restore.html [hunk 3]
```diff
@@ -0,0 +1,26 @@
+<div ng-controller="Umbraco.Editors.Media.RestoreController">
+	<div class="umb-dialog-body">
+		<umb-pane>
+
+			<p class="abstract" ng-hide="error != null || success == true">
+				Restore <strong>{{currentNode.name}}</strong> to under <strong>{{target.name}}</strong>?
```
Comment (dawoe) [inline src/Umbraco.Web.UI.Client/src/views/media/restore.html]: The texts in this view should be retreived from the language files using the localize directive

**Comment by richarth** [comment]
Hi,

I've fixed the project file.

Is there a good example that I could follow for the localisation?  I'm happy to do the same localisation changes for content if I can get some pointers in the right direction.

Thanks

Richard

**Comment by dawoe** [comment]
Hi @richarth 

You can see some examples of doing localization in this PR : https://github.com/umbraco/Umbraco-CMS/pull/2159. It contains code in the api controllers for getting translation as well in angular controllers and views. If you have any questions regarding this please let me know.

I also tested it a bit already. It works great on a new 7.12.0 install.  For an upgrade I needed to fix the migration. After that it works also. So good job !!

Here you can find a gist with the updated migration : https://gist.github.com/dawoe/283fc6867a197fccfc4b31670d587666

Dave

**Comment by richarth** [comment]
Hi Dave,

I've had a go at the localisation.  I'd be grateful if you could review please and let me know if you have any feedback.

Richard

**Comment by dawoe** [comment]
Hi @richarth 

I just got back from holidays and am ready to review this one again.

Could you make sure the PR is mergeable ?

Dave

**Comment by nul800sebastiaan** [comment]
@dawoe This is a line ending problem, see https://www.zpqrtbnk.net/posts/git-line-endings

For reviewing purposes you can ignore this for now, we'll fix it manually before this gets merged! 👍

**Comment by dawoe** [comment]
@nul800sebastiaan i don't think it's line ending

When i pull it locally and merge in latest dev-v7 the Umbraco.Core project won't load in VS.

Dave

**Comment by nul800sebastiaan** [comment]
Correct. I'll fix it, one minute.

**Comment by nul800sebastiaan** [comment]
@dawoe Actually.. no, don't know what you're doing but it loads for me.

And [if you look here](https://github.com/umbraco/Umbraco-CMS/pull/2342/files?w=1#diff-68048b09ab33c708c240fc3c3c3cbbff), with `?w=1` added to the URL (that ignores whitespace changes in the diff), you can see it's just adding a file and 

image

**Comment by nul800sebastiaan** [comment]
I've resolved the merge conflict now, maybe that helps for you?

**Comment by nul800sebastiaan** [comment]
Awesome! I've also just tested it, looks good on both upgrades and new installs, works as expected and the code looks good!

Thanks very much @richarth - great job! ⭐🏅

**Pull request merged**